groups/ntc/ntcq/ntcq_zerocopy.cpp

```diff
@@ -19,187 +19,453 @@
+
+    *dataContainer = data;
+
+    d_waitList.resize(d_waitList.size() + 1);
+    ZeroCopyEntry& entry = d_waitList.back();
+
+    entry.setGroup(group);
+    entry.setMinCounter(counter);
+    entry.setMaxCounter(counter + 1);
+    entry.setData(dataContainer);
+
+    if (callback) {
+        entry.setCallback(callback);
     }
-    d_callbacks.clear();
+
+    return counter;
 }
 
-ZeroCopyWaitList::ZeroCopyWaitList(bslma::Allocator* basicAllocator)
-: d_entryList(basicAllocator)
-, d_strand()
-, d_nextId(0)
+ntcq::ZeroCopyCounter ZeroCopyQueue::push(
+    ntcq::SendCounter                  group,
+    const bsl::shared_ptr<ntsa::Data>& data)
 {
+    BSLS_ASSERT(d_waitList.empty() || d_waitList.front().group() < group);
+
+    ntcq::ZeroCopyCounter counter = d_generator.next();
+
+    d_waitList.resize(d_waitList.size() + 1);
+    ZeroCopyEntry& entry = d_waitList.back();
+
+    entry.setGroup(group);
+    entry.setMinCounter(counter);
+    entry.setMaxCounter(counter + 1);
+    entry.setData(data);
+
+    return counter;
 }
 
-ZeroCopyWaitList::~ZeroCopyWaitList()
+ntcq::ZeroCopyCounter ZeroCopyQueue::push(
+    ntcq::SendCounter                  group,
+    const bsl::shared_ptr<ntsa::Data>& data,
+    const ntci::SendCallback&          callback)
 {
-    BSLS_ASSERT(d_entryList.empty());
+    BSLS_ASSERT(d_waitList.empty() || d_waitList.front().group() < group);
+
+    ntcq::ZeroCopyCounter counter = d_generator.next();
+
+    d_waitList.resize(d_waitList.size() + 1);
+    ZeroCopyEntry& entry = d_waitList.back();
+
+    entry.setGroup(group);
+    entry.setMinCounter(counter);
+    entry.setMaxCounter(counter + 1);
+    entry.setData(data);
+
+    if (callback) {
+        entry.setCallback(callback);
+    }
+
+    return counter;
 }
 
-void ZeroCopyWaitList::setStrand(const bsl::shared_ptr<ntci::Strand>& strand)
+ntcq::ZeroCopyCounter ZeroCopyQueue::push(ntcq::SendCounter group)
 {
-    d_strand = strand;
+    NTCCFG_WARNING_UNUSED(group);
+
+    BSLS_ASSERT(!d_waitList.empty());
+
+    ntcq::ZeroCopyCounter counter = d_generator.next();
+
+    ZeroCopyEntry& entry = d_waitList.back();
+    BSLS_ASSERT(entry.group() == group);
+
+    entry.setMaxCounter(counter + 1);
+
+    return counter;
 }
 
-void ZeroCopyWaitList::addEntry(ZeroCopyEntry& entry)
+void ZeroCopyQueue::frame(ntcq::SendCounter group)
 {
-    BSLS_ASSERT(!cancelled);
+    NTCCFG_WARNING_UNUSED(group);
+
+    ZeroCopyEntry& entry = d_waitList.back();
+    BSLS_ASSERT(entry.group() == group);
 
-    entry.setId(d_nextId++);
+    entry.setFramed(true);
 
-    d_entryList.push_back(entry);
+    if (entry.complete()) {
+        d_doneList.push_back(entry);
+        d_waitList.pop_back();
+    }
 }
 
-bool ZeroCopyWaitList::zeroCopyAcknowledge(
-    const ntsa::ZeroCopy&                  zc,
-    const bsl::shared_ptr<ntci::Sender>&   sender,
-    const bsl::shared_ptr<ntci::Executor>& executor)
+ntsa::Error ZeroCopyQueue::update(const ntsa::ZeroCopy& zeroCopy)
```

Comment:
It looks like this method wil not handle correctly the situation when we are receiveing a ZC notification for pre-overflowing period.
let's imagine that the kernel maintains uint8_t ZC counter.
```
kenel counter goes this way: ..... 252 253 254 255   0   1   2   3  4    5
our ntcq::ZeroCopyCounter: .....   252 253 254 255 256 257 258 259 260 270
```
Then imagine we receive ZC notification [255, 1) -> we'd set generation to 1 and bias to 255 (uint8 in this example)
The next ZC notification is [250, 254) -> we may incorrectly calculate offset in this case.

Probably we also need to take into account a distance from the last overflow moment  (just thinking loud)

Reply:
I also wondered about this and I assume the kernel has to delay and split announcement of ZC wraparound, otherwise I've know idea offhand how we are supposed to handle this.

Reply:
Let's postpone solution for that

Reply:
Disregard my previous comment. There's little likelihood the Linux kernel would decide to do anything that helps us in this case. This PR has a bug that it doesn't handle this case and will mispresent out-of-order intervals after wrap-around has being wrapped-around (part of the next generation) when they are really part of the last generation.